crates/iota-indexer/src/db.rs

```diff
@@ -229,61 +241,101 @@ pub mod setup_postgres {
         diesel::sql_query(drop_all_functions).execute(conn)?;
         info!("Dropped all functions.");
 
-        diesel::sql_query(
-            "
-        CREATE TABLE IF NOT EXISTS __diesel_schema_migrations (
-            version VARCHAR(50) PRIMARY KEY,
-            run_on TIMESTAMP NOT NULL DEFAULT NOW()
-        )",
-        )
-        .execute(conn)?;
+        conn.setup()?;
         info!("Created __diesel_schema_migrations table.");
 
-        conn.run_migrations(&MIGRATIONS.migrations().unwrap())
-            .map_err(|e| anyhow!("Failed to run migrations {e}"))?;
+        run_migrations(conn)?;
         info!("Reset database complete.");
         Ok(())
     }
 
+    /// Execute all unapplied migrations.
+    pub fn run_migrations(conn: &mut PoolConnection) -> Result<(), anyhow::Error> {
+        conn.run_pending_migrations(MIGRATIONS)
+            .map_err(|e| anyhow!("Failed to run migrations {e}"))?;
+        Ok(())
+    }
+
+    /// Checks that the local migration scripts is a prefix of the records in
+    /// the database. This allows us run migration scripts against a DB at
+    /// anytime, without worrying about existing readers fail over.
+    /// We do however need to make sure that whenever we are deploying a new
+    /// version of either reader or writer, we must first run migration
+    /// scripts to ensure that there is not more local scripts than in the DB
+    /// record.
```

Comment:
I would also add that this requires that new migrations are backward compatible with respect to the model.

Reply:
https://github.com/iotaledger/iota/pull/7146/commits/0e8a4ff76cf1ae5bc72361b1c8a19cdedb5743fa

Reply:
https://github.com/iotaledger/iota/pull/7146/commits/fbf2ac857869957e304a80399ca6ff14640ee970